### ISSUE TITLE
[rl] Observations unvalidated at replay ingestion — the unguarded `act` path (no loss, no guard) is the live half; C51 learn-path NaN is caught by FiniteLossGuard on both backends

### ISSUE THREAD

**Issue by anthonytorlucci** [anthonytorlucci/rlevo#1043]
Surfaced while resolving #352 and named as explicitly out of scope in **ADR 0065**, but never filed. Filing it so it is not lost the way C51 and QR-DQN's `remember` holes were lost between the four sites #352 named and the six that actually existed.

## The gap

ADR 0065 closed reward finiteness at `remember`. The **observations passing through the same call, in the same struct literal, on the same line**, are still unvalidated:

```rust
// crates/rlevo-reinforcement-learning/src/algorithms/dqn/dqn_agent.rs:386
pub fn remember(&mut self, obs: O, action: &A, reward: f32, next_obs: O, terminated: bool) {
    if !self.reward_guard.admit(reward) {   // <- guarded
        return;
    }
    self.buffer.push(DiscreteTransition {
        obs,                                 // <- not guarded
        action: action.to_index(),
        reward,
        next_obs,                            // <- not guarded
        terminated,
    });
}
```

Same shape in `ddpg_agent.rs:417`, `td3_agent.rs:466`, `sac_agent.rs:476`, `c51_agent.rs:405`, `qrdqn_agent.rs:366`.

A non-finite element in `obs`/`next_obs` flows into staging at `dqn_agent.rs:534-535` (`t.obs.write_host_row(&mut obs_flat)` / `t.next_obs.write_host_row(&mut next_flat)`), into the observation tensor at `:546`, and straight into the network's forward pass. Nothing between the environment and the first matmul checks it. The same is true on the `act` path (`dqn_agent.rs:322,352`, `obs.to_tensor(&self.device)`).

## Why it is not simply "#352 again"

The failure mode is worse in one respect and the fix is more expensive in another.

**Worse:** a NaN reward corrupts one scalar in the Bellman target. A NaN *observation* is fed through the network's weights, so a single poisoned element propagates across the entire output layer within one forward pass — and on the `act` path it does so with no loss guard downstream at all, because `act` performs no `backward()`. ADR 0056's `FiniteLossGuard` only protects the learn path; a NaN observation at action-selection time produces a NaN action with nothing to catch it. Worth checking whether that then reaches the environment's `step` and what the environments do with it (`clamp` will propagate the NaN, per `c51/projection.rs:97`).

**More expensive:** the reward check is one register compare on an `f32` already in hand. An observation check is a **per-element scan** over `O`, whose size is unbounded — `pixel_grid` and `CarRacing` observations are image-shaped. That is a genuinely different cost/benefit, which is exactly why ADR 0065 scoped it out rather than folding it in. It is the reason this needs its own decision, not a reason to skip it.

## The design questions this needs answered

- **Where.** `remember` only (learn path), or also `act` (which has no downstream guard at all and is arguably the more urgent half)?
- **Cost.** Is a per-element scan acceptable per step? `Observation`/`HostRow<R>` may allow the check to ride the existing `write_host_row` traversal at near-zero marginal cost — the row is already being walked. That would make the learn-path check nearly free and is worth measuring before assuming it is expensive.
- **Release vs debug.** ADR 0065 and ADR 0056 both run unconditionally in release, on the reasoning that a gated guard is stripped from exactly the long diverging runs that need it. That reasoning was cheap to accept when the check was a register compare. It must be re-argued here, not inherited.
- **Response.** Drop-and-count (ADR 0065's semantic) transfers cleanly for `remember`. It does **not** transfer to `act` — there is no transition to drop, and the agent must still return *some* action.

## Acceptance

- A decision recorded in an ADR, cross-referencing ADR 0065 and stating why the reward's semantic does or does not transfer to each of the two paths.
- A benchmark backing the cost claim, whichever way it lands — the cheap-vs-expensive question is the crux and should not be settled by assertion.
- A regression test with a deliberately NaN-emitting observation.

## Related

- #352 / ADR 0065 — the reward half, resolved; names this as out of scope
- #318 / ADR 0056 — `FiniteLossGuard`, which does not cover the `act` path at all
- #1042 — the on-policy PPO/PPG rollout reward hole

**Comment by anthonytorlucci** [comment]
**The C51-specific mechanism this issue was cross-referenced against — a NaN observation reaching the categorical projection's Bellman-shift coordinate — does not exist. A reviewer who re-checks that specific claim against the code will see it disagree, so stating this plainly before anyone closes this on the strength of the general framing alone:** `tz` and `b` in `project_distribution` are derived **only** from `rewards`, `terminated`, and the fixed `support` tensor, never from network output.

```rust
// crates/rlevo-reinforcement-learning/src/algorithms/c51/projection.rs:182-188
let rewards_bn: Tensor<B, 2> = rewards.unsqueeze_dim::<2>(1);
let terminated_bn: Tensor<B, 2> = terminated.unsqueeze_dim::<2>(1);
let support_bn: Tensor<B, 2> = support.unsqueeze_dim::<2>(0);
let keep = terminated_bn.neg().add_scalar(1.0);
let tz = rewards_bn + keep * support_bn * gamma;
```
```rust
// projection.rs:235
let b = (tz.sub_scalar(v_min)).div_scalar(delta_z);
```

Neither line touches `next_probs`. The target network's forward pass over `next_obs` only enters `project_distribution` as the `next_probs` argument (`c51_agent.rs:594,603-607`, softmax over `next_logits_inner`), which multiplies `weight_lower`/`weight_upper` (`projection.rs:249-250`) — a *value* scattered into the target, never the coordinate that derives the scatter *index*. So #1044's own reachability section, which named this issue as "the most plausible remaining route to a NaN `b`," was wrong about the route: a NaN observation cannot reach `b` at all, by construction, and there is no backend-divergent clamp in its actual path.

## What a NaN observation actually does on the C51 learn path

`softmax` of a row containing one NaN logit gives an **all-NaN row on every backend**, no rescue anywhere, and this doesn't turn on any clamp: `B::softmax` is `exp(x - max_dim(x)) / sum_dim(exp(...))` (`burn-backend-0.21.0/src/backend/ops/activation.rs:250-256`). Whatever the internal `max_dim` reduction does with the NaN entry, the `sum_dim` step is ordinary addition — there is no NaN-suppressing hardware variant of `+` the way `min`/`max` have `minNum`/`maxNum` — so the sum is NaN and the final division poisons every element of the row identically on Flex and wgpu.

That NaN then: enters only through `next_probs_chosen` → `weight_lower`/`weight_upper` → the scattered target (`projection.rs:249-254`); flows through `categorical_cross_entropy_per_sample` (`c51/loss.rs:37-45`, `sum_dim` again, no clamp anywhere in the function); and into the mean reduction. The batch loss is NaN, and `c51_agent.rs`'s `loss_guard.check(loss_value)` (`:677`) fires — **identically on both backends**, because nothing on this path is a clamp with backend-divergent NaN semantics. This is the correct, working case; ADR 0056's guard was never actually at risk here.

## What survives — and should become this issue's primary claim

The half of this issue's own framing that holds up, and the half its body already flagged as "arguably the more urgent half," is now the *only* half: the `act` path has no loss, no `backward()`, and no guard of any kind.

```rust
// c51_agent.rs:320-329 (act_greedy)
let probs: Tensor<B, 3> = activation::softmax(logits, 2);
...
let q: Tensor<B, 2> = (probs * support_3d).sum_dim(2).squeeze_dim::<2>(2);
let idx = q.argmax(1).into_scalar();
A::from_index(idx.elem::<i64>() as usize)
```

A NaN observation poisons `probs` (same all-NaN-row softmax mechanism as above), `q` is all-NaN, and `argmax` over an all-NaN row returns *some* index with zero diagnostic on either backend — no warning, no error, just a silently arbitrary action. `act_greedy_with` (`c51_agent.rs:363-377`) is an identical-shape sibling with the same gap (it exists specifically for the hot inference path, so it is not a lower-traffic near-miss).

## Recommendation

- Retitle to name the `act`-path gap specifically, since the learn-path half of the current title/framing is refuted for C51 (the DQN/DDPG/TD3/SAC/QR-DQN `remember`-validation gap this issue is actually about is unaffected by any of the above — this comment only concerns the C51-specific reachability trace inherited from #1044).
- Checked current labels: only `bug` is present, no `priority: high` label to strip. If a priority label is added later, it should be earned by the `act`-path gap (which is real and unguarded on every agent, not just C51), not by the now-refuted C51 learn-path trace.
- Leaving this open — the underlying `remember`/`act` observation-validation gap is still real; only the specific C51-projection mechanism named via #1044 is refuted.
